enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+processes have to each backend server.  This lack of data can cause uneven
+traffic load over backend servers from the aggregate set of haproxy processes
+because the new process balances over the set of backend services without any
+coordination with the old processes.
+
+Dynamic Config Manager addresses this issue again by eliminating the need to
+reload the configuration for endpoints changes.  Because adding and removing
+servers is possible to do through HAProxy's control socket, DCM is able to
+eliminate the problem of uneven load resulting from adding or removing
+endpoints.
+
+Note that DCM is not able to prevent imbalance in the event of an endpoints
+update combined with an update that requires a configuration reload.  DCM also
+cannot prevent imbalance in the case of many router pod replicas with relatively
+few connections.  For example, multiple router pod replicas each receiving a
+single request for the same route can all choose the same backend server; router
+pod replicas do not coordinate with each other.
+
+### Goals
+
+- DCM is enabled by default on OpenShift 4.18 clusters.
+- Updating a route's endpoints does not trigger an HAProxy configuration reload.
+  - Initially, this might be true only for a single scale-out event, or subsequent scale-in event, per route.
+- OpenShift router uses marginally more memory and CPU with DCM than without.
+- If DCM cannot handle an update, the router forks and reloads, same as before.
+- DCM does not cause regressions in throughput, latency, balancing, or metrics.
+
+### Non-Goals
+
+- DCM cannot handle *all* route or endpoints configuration changes.
+  - Initially, adding and removing backends will still require a reload.
+  - Initially, changing certificates or annotations will still require a reload.
+  - Initially, multiple successive scale-out events will still require a reload.
+- DCM cannot handle *router* configuration changes (that is, *global* options).
+  - DCM does not handle the `timeout`, `maxconn`, `nbthread`, or `log` options.
+  - Updates to the router configuration still generally require a pod restart.
+- DCM does not coordinate among router pods.
+  - Traffic can still be unbalanced if multiple router pods use the same server.
+
+## Proposal
+
+OpenShift router's Dynamic Config Manager (DCM) was implemented in OpenShift
+Enterprise 3.11 with HAProxy 1.8.  Although DCM was not previously enabled in
+OpenShift Container Platform 4, the implementation was never removed from the
+source code.  However, it has not been actively developed by engineering or
+tested by QA or in CI in OpenShift 4.  Therefore, this enhancement proposes the
+following steps:
+
+1. Manually verify that the router functions and passes E2E tests with DCM enabled.
+2. Add a tech-preview featuregate in openshift/api for DCM.
+3. Update cluster-ingress-operator to enable DCM if the featuregate is enabled.
+4. Re-enable old E2E tests in openshift/origin for DCM.
```

Comment:
Can this be done if DCM will be in techpreview featureset?

Reply:
Yes.  The openshift/origin tests create router pods directly, so the tests can enable DCM irrespective of the featuregate: https://github.com/openshift/origin/blob/203d2b124037a543d8719ba7e2a369f52f01686d/test/extended/router/config_manager.go#L328-L489